### DISCUSSION TITLE
AIOHTTP Backend Support RFC

### DISCUSSION THREAD

**Discussion by bitnom** [idom-team/idom#939] [Ideas]
[AIOHTTP](https://github.com/aio-libs/aiohttp) is (IMO) a wonderful backend. Much performance can be gained with it, especially when combined with `ProcessPoolExecutor` and `SO_REUSEPORT` (Websockets supported), and there's a wealth of modules to extend it. Has there yet been any effort to add support for it, official or otherwise? I skimmed through [django-idom](https://github.com/idom-team/django-idom) but there's a lot going on there I'm unfamiliar with. I rarely use DJango unless I have to, much less go under-the-hood.

**Comment by rmorshea** [comment]
There has not. I wouldn't be turned off by the idea of creating a backend for AIOHTTP based on django-idom though. django-idom is much more complicated because Django itself is way more than just a web server. Examples of existing integrations distributed with the core IDOM library can be found [here](https://github.com/reactive-python/reactpy/tree/main/src/reactpy/backend). Hopefully you'll find that these built-in backends are much simpler to implement. A spec for the routes backends must expose has not been documented, but the gist is that backends should have the following routes (copied from [this PR](https://github.com/idom-team/idom/pull/824)):

```
/_idom/assets/<file-path>
/_idom/stream/<spa-path>
/_idom/modules/<file-path>
/<spa-path>
```

- `/_idom/assets` - where files found in the `assets` folder of the built client under [`idom.backend._common.CLIENT_BUILD_DIR`](https://github.com/idom-team/idom/blob/main/src/idom/backend/_common.py#L25)
- `/_idom/stream` - websockets for the associated `spa-path`
- `/_idom/modules` - hosts web modules that can be dynamically loaded from [`idom.config.IDOM_WEB_MODULES_DIR`](https://github.com/idom-team/idom/blob/main/src/idom/config.py#L44)
- `/<spa-path>` - the actual single-page-app (note that the client will use the same `spa-path` when connecting to a stream)

**Reply by rmorshea** [reply]
Just note that we're [renaming this project to ReactPy](https://github.com/idom-team/idom/discussions/882) and wont be able to accept any contributions until that's complete. Thankfully we're working on this [right now](https://github.com/idom-team/idom/pull/892) and I expect that we'll have that done in the next week or two.

**Reply by bitnom** [reply]
Thanks for the info. That link is 404 btw. Looks like it's now [here](https://github.com/reactive-python/reactpy/tree/main/src/reactpy/backend). I may take a crack at a backend.

**Comment by Archmonger** [comment]
Right now, Django-IDOM is the most feature complete way of using IDOM. Django might get a bit overwhelming due to built in support for database, cache, HTML templating and rendering, etc.

This [9 minute video](https://www.youtube.com/watch?v=ZsJRXS_vrw0) summarize it pretty well. It's an extremely opinionated framework where most code must be written a certain way.

On the note of other ASGI frameworks though, long term we'd like contributions to support all major ASGI frameworks. I've been eyeing `Blacksheep` support for a while due to [amazing performance benchmarks](https://klen.github.io/py-frameworks-bench/).

**Reply by bitnom** [reply]
wow there's a lot more built-in to blacksheep than I had realized.

As for Django, I just know I'd never start new projects with it. If it had competitive performance, I might change my mind. Otherwise, it might haunt my dreams. After having finally just gone through all the blacksheep docs, I won't be able to resist it.

**Reply by Archmonger** [reply]
Getting blacksheep working with IDOM should be pretty simple since it follows a flask-like API.

You can clone the IDOM `backend.flask` module and use that as a base.

**Reply by Archmonger** [reply]
Related note, Django ASGI performance is currently severely hurt by the fact that they made the decision to prioritize WSGI performance at all costs... _Kinda ironic since Django actually wrote the ASGI spec._ So right now ASGI in Django is more of a gimmick than a feature.

I've been planning on developing an `ASYNC_ONLY` mode in Django that inverts that decision. Whether that be PR'd upstream or an external monkey-patch package is yet to be seen. But, a lot of that design will be made in inspiration of blacksheep with the goal of providing modern ASGI performance to Django.
